pallets/file-system/src/utils.rs

```diff
@@ -176,10 +206,91 @@ where
         Ok(())
     }
 
+    pub(crate) fn do_bsp_confirm_storing(
+        who: StorageProviderId<T>,
+        location: FileLocation<T>,
+        root: FileKey<T>,
+        proof: Proof<T>,
+    ) -> DispatchResult {
+        let bsp = match <T::Providers as storage_hub_traits::ReadProvidersInterface>::get_provider(
+            who.clone(),
+        ) {
+            Some(bsp) => bsp,
+            None => return Err(Error::<T>::NotABsp.into()),
+        };
+
+        // Check that the storage request exists.
+        let file_metadata = expect_or_err!(
+            <StorageRequests<T>>::get(&location),
+            "Storage request should exist",
+            Error::<T>::StorageRequestNotFound
+        );
+
+        expect_or_err!(
+                    file_metadata.bsps_confirmed < file_metadata.bsps_required,
+                    "Storage request should never have confirmed bsps equal to or greater than required bsps, since they are deleted when it is reached.",
+                    Error::<T>::StorageRequestBspsRequiredFulfilled,
+                    bool
+                );
+
+        // Check that the sender is a registered storage provider.
+        ensure!(
+            <StorageRequestBsps<T>>::contains_key(&location, &who),
+            Error::<T>::BspNotVolunteered
+        );
+
+        // Check that the storage provider has not already confirmed storing the file.
+        ensure!(
+            !<StorageRequestBsps<T>>::get(&location, &who)
+                .expect("BSP should exist since we checked it above")
+                .confirmed,
+            Error::<T>::BspAlreadyConfirmed
+        );
+
+        // Check that the proof is valid.
+        ensure!(
+            <T::ProofDealer as storage_hub_traits::ProofsDealerInterface>::verify_proof(
+                &bsp, &root, &proof,
+            )
+            .is_ok(),
+            Error::<T>::InvalidProof
+        );
+
+        // Increment the number of confirmed storage providers.
+        <StorageRequests<T>>::try_mutate(&location, |file_metadata| -> DispatchResult {
+            let file_metadata = file_metadata
+                .as_mut()
+                .ok_or(Error::<T>::StorageRequestNotFound)?;
+
+            match file_metadata
+                .bsps_confirmed
+                .checked_add(&T::StorageRequestBspsRequiredType::one())
```

Comment:
Shouldn't we check if there are already enough BSPs confirmed at the beginning? Before doing heavy comptation like `verify_proof`?

Reply:
yes!